### DISCUSSION TITLE
How do I implement debug mode?

### DISCUSSION THREAD

**Discussion by Zachary-Rude** [Techokami/SonicWorldsNext#158] [Q&A] [answered]
Hi, I'm working on a game using this engine, and I would like to implement the debug mode from the classic Sonic games, where you can press a specific button during a level (usually after entering a button combination on the title screen or a specific set of numbers on the sound test) and you will be able to place various objects and freely fly around the level, with coordinates being shown either in place of the score counter or in the top right corner. Here's some screenshots of the debug mode in various classic Sonic games:
![Sonic 1 debug mode](https://github.com/Techokami/SonicWorldsNext/assets/84681078/f9503d62-3dc6-4cfc-8b77-8840f12d986b)
![Sonic 2 debug mode](https://github.com/Techokami/SonicWorldsNext/assets/84681078/0dd913d9-9c46-4a2c-8205-47d5b9cc7b37)
![Sonic 3 & Knuckles debug mode](https://github.com/Techokami/SonicWorldsNext/assets/84681078/1daac467-7ec1-4782-98d5-925683ed7af6)
![Sonic Mania debug mode](https://github.com/Techokami/SonicWorldsNext/assets/84681078/06c48093-c5cc-4b44-bccc-93065e7bd024)

I'm not necessarily asking for you to give me code, I would just like you to tell me what the code _should_ do so that I can write the code by myself. I'm glad to be using this engine!

**Comment by DimensionWarped** [comment, marked as answer]
First thing you should probably try to implement is the four directional flying behavior.

First thing I'd do... add some piece of script somewhere that gets skipped if debug mode is disabled. Maybe even just in global.gd

In here, if player 1's debug mode button is pushed, you start the process of going into/leaving debug mode.

Add a new player state for debug. When you enter this state, disable all the player's collision. In the player state code, you'll also have your new movement controls (which are very simple and just involve a slow but steady acceleration to the direction you are currently holding -- you can fine tune that however you want).

Also as part of entering debug mode, you'll enable the display of your coordinates... along with any other values you think are important. Don't feel obligated to actually display them in the same way as classic Sonic. This is a modern engine, you don't need to carefully repurpose things like you would have if you were writing for a Motorola 68000 or whatever. In fact, I'd probably just add a new UI element in the top right. I understand if authenticity is important to you, but I personally feel like authenticity is for features that are part of the core experience, not for debug tools.

Obviously when you leave debug mode you'll set the player's state to normal, you'll re-enable display of the player's sprite (if you've done that part... I wouldn't disable/enable the player sprite myself until I had started work on the object painting stuff).

Speaking of the object painting stuff, the basic approach I would take to this right now would be to give the level object (the same one that currently controls which music plays in a scene along with what the next act is) a customizable list of objects that can be spawned. This would be a simple array like the following:

{{graphic1, object to spawn1},
 {graphic2, object to spawn2},
 {graphic3, object to spawn3},
 ...
}

In your debug code back in global, add counter to track where in the array we are. When we enter debug mode, disable the display of the first player's sprite and enable the display of our object painter which will always just draw the graphic of the currently tracked object in that list. When the player pushes their A equivalent button while in debug mode, increment the counter by one (and roll over if the counter is equal to the length of the array above). When the player presses their C equivalent button, spawn the object in.

It's very possible I'm missing some things here, but that should cover the basics I think.

**Reply by Zachary-Rude** [reply]
I'll try that later!

**Comment by Zachary-Rude** [comment]
Here's the result that I got, what do you think of it?

https://github.com/Techokami/SonicWorldsNext/assets/84681078/a5fdf070-c5e1-401a-ada9-00d3eb030998

I added a cheat code that needs to be entered on the title screen in order for debug mode to be unlocked, and the ring sound will play once the code has successfully been inputted, just like in the classic Sonic games.

**Reply by Renhoex** [reply]
hey this looks pretty spot on, good job

**Reply by Zachary-Rude** [reply]
Thanks!

**Reply by Raptorq1** [reply]
Hey man do you have the code for debug mode and for the cheat code because I tried implementing it my self but it didn't work

**Reply by Zachary-Rude** [reply]
@Raptorq1 I can't give you the actual code that I wrote, but if it didn't work, then you should probably try either following @DimensionWarped's instructions better or experimenting with possible solutions to fix your problem. Speaking of which, what exactly _is_ your problem, anyway? Like, which aspect of the debug mode isn't working?
And as for the cheat code implementation, what I did is to check if debug mode hasn't already been enabled, and if so, then check if a specific button has been pressed, and if it hasn't already been pressed more than a specific amount of times, then we will increment a counter, and we will repeat this for all the buttons that will be used within the cheat code, and after all of the correct buttons have been pressed in the correct order, then we will enable debug mode (what I did for that is to use a variable within `Global.gd` that is false by default, and set that variable to true in order to enable debug mode) and play a ring sound (you will need to add an `AudioStreamPlayer` node to the title scene and use the audio file for the ring sound effect for the node's stream property) to let the player know that the code was correctly entered.
I could be missing a few things, but that should give you a basic idea of what I did to add the button combination cheat code to the title screen.
(sorry for the wall of text)

**Reply by DimensionWarped** [reply]
Looks pretty badass!

EDIT: Out of curiosity, how much of my advice did you end up taking? There are a lot of ways to crack this egg I imagine.

**Reply by Zachary-Rude** [reply]
@DimensionWarped Thanks! I took most of your advice, but I decided to put the debug code for the counter and all that stuff within the debug player state in order to make things easier for me, and I put a Boolean variable in `Global.gd` for enabling and disabling debug mode, and this variable is what all the other scripts use to check if debug mode is enabled, it's set to false by default but once the cheat code has been entered on the title screen, the variable is set to true.

**Reply by Raptorq1** [reply]
> @DimensionWarped Thanks! I took most of your advice, but I decided to put the debug code for the counter and all that stuff within the debug player state in order to make things easier for me, and I put a Boolean variable in `Global.gd` for enabling and disabling debug mode, and this variable is what all the other scripts use to check if debug mode is enabled, it's set to false by default but once the cheat code has been entered on the title screen, the variable is set to true.

Could you help me because when i hidethe player it doesnt 


```
extends PlayerState


func _process(delta: float) -> void:
	hide_player()

func hide_player():
	if Global.debug == 1:
		if Input.is_action_just_pressed("gm_action3"):
			print("debug mode enabled")
			parent.get_node("../Sonic").visible = false
			parent.get_node("HitBox").disabled = true
	else:
		print("debug mode disabled")
		parent.get_node("Sonic").visible = true
		parent.get_node("HitBox").disabled = false
```
